### PR TITLE
Fixing Sync Notifications Race Condition

### LINKED ISSUE #6213: Notifications: Duplicate Notifications on 6.7
#Whilst using the latest 6.7 release over the weekend, I noticed that my Notifications section included a number of duplicated notifications: two notifications (one mention, one '10 likes') had repeated 3x each:

img_4947

I viewed one of the mention comments at the bottom, and the app then marked **two** of them as being read:

img_4948

I was using my iPhone 7, which is on the latest version of iOS.

cc @jleandroperez

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details:
This PR introduces a privately-shared `Derived Context`, to be used by all of the `NotificationSyncService` instances.

Reason for doing so is: 
- Multiple `sync` calls may get triggered before any of them finishes (ie. many Push Notifications arriving nearly at the same time)
- All of the Sync'ing code works on a `derived context`
- The work (Insert / Update / Delete) OP's is merged back into the Main MOC on completion
- There may be multiple workers unaware of whatever the other is doing
- By sharing a common DerivedContext we simply neutralize any and all of the race conditions.

Fixes #6213

### Testing:
- New Unit Test added: `testMultipleSyncCallsWontInsertDuplicateNotes` (NotificationSyncServiceTests)
- Please, review the changes, run the test, and make sure everything is happiness.

Needs review: @astralbodies 
Sir, may i bug you with a review?

Thanks!

cc @frosty (Thanks James for filing the issue!!)

### CODE REVIEW

File: WordPress/Classes/Services/NotificationSyncService.swift [hunk 1]
```diff
@@ -297,3 +305,29 @@ private extension NotificationSyncService
...
     }
 }
+
+
+
+// MARK: - Thread Safety Helpers
+//
+extension NotificationSyncService
+{
+    /// Returns the current Shared Derived Context, if any. Otherwise, proceeds to create a new
```
Comment (astralbodies) [inline WordPress/Classes/Services/NotificationSyncService.swift]: Trailing whitespace violation here...

File: WordPress/Classes/Services/NotificationSyncService.swift [hunk 2]
```diff
@@ -22,12 +22,20 @@ class NotificationSyncService
...
         return contextManager.mainContext
     }
 
+    /// Thread Safety Helper!
+    ///
+    private static let lock = NSLock()
+
+    /// Shared PrivateContext among all of the Sync Service Instances
+    ///
+    private static var privateContext: NSManagedObjectContext!
```
Comment (astralbodies) [inline WordPress/Classes/Services/NotificationSyncService.swift]: Ughhhh this feels so nasty. One of the founding principles of the service architecture in the app is no state is retained between method calls.

**Comment by jleandroperez** [comment]
@astralbodies sir, ready for another look. As ever, thanks for helping me out!!!

**Comment by jleandroperez** [comment]
Much appreciated @astralbodies!!!

**Pull request merged**